url46a88cc932_toby1984_tinyc_tgz-pJ8-de_codesourcery_dcpu16_compiler_CompilerJ8/src/de/codesourcery/dcpu16/optimizer/cfg/ControlFlowGraph.java

```diff
@@ -0,0 +1,389 @@
+                    allBlocks.remove( block );
+                    
+                    predecessor.blockRemoved( block );
+                    for ( BasicBlock successor : successors ) {
+                        successor.blockRemoved( block );
+                        predecessor.addSuccessor( successor );                        
+                    }
+                    cont = true;
+                    break;
+                }
+            }
+        } while( cont );
+    }
+
+    public  static final class Edge 
+    {
+        public final BasicBlock from;
+        public final BasicBlock to;
+
+        public Edge(BasicBlock from, BasicBlock to)
+        {
+            this.from = from;
+            this.to = to;
+        }
+
+        @Override
+        public int hashCode()
+        {
+            int result = 31 + from.hashCode();
+            result = 31 * result + to.hashCode();
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            if (this == obj) {
+                return true;
+            }
+            if ( !(obj instanceof Edge) ) {
+                return false;
+            }
+            final Edge other = (Edge) obj;
+            return this.from.equals( other.from ) && this.to.equals(other.to);
+        }
+    }
+    public class BasicBlock 
+    {
+        private final String identifier;
+        private final Set<Edge> edges = new HashSet<>();
+
+        private final Map<String,Object> metaData = new HashMap<>();
+        private final List<ASTNode> astNodes = new ArrayList<>();
+
+        public BasicBlock() {
+            identifier="block_"+nodeId;
+            nodeId++;
+        }
+        
+        public boolean hasSuccessor(BasicBlock bb) {
+            return getSuccessors().contains(bb);
+        }
+        public Map<String, Object> getMetaData()
+        {
+            return metaData;
+        }
+
+        public void blockRemoved(BasicBlock bb)
+        {
+            for (Iterator<Edge> it = edges.iterator(); it.hasNext();) {
+                Edge e = it.next();
+                if ( e.from.equals( bb ) || e.to.equals( bb ) ) {
+                    it.remove();
+                }
+            }
+        }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            return obj instanceof BasicBlock && ((BasicBlock) obj).identifier.equals( this.identifier );
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return identifier.hashCode();
+        }
+
+        public BasicBlock(String identifier) {
+            this.identifier = identifier;
+        }
+
+        public Set<BasicBlock> getSuccessors() 
+        {
+            Set<BasicBlock> result = new HashSet<>();
+            for ( Edge e : edges ) {
+                if ( e.from.equals( this ) ) {
+                    result.add(e.to);
+                }
+            }
+            return result;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 231:237, 242:248.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.